feat: introduce extended bank transaction fields

0xD0M1M0 opened the pull request:
Bank Transactions, e.g. to a foreign currency, often come with a fee. This is often listed as an additional column in the bank statement. 

This PR introduces the generic fee field to allow the import and future apps to handle fees if necessary.

In included_fee is inside the withdrawal or deposit amount, while the excluded_fee comes with some banks in a separate column. This way the bank import can handle both types. The excluded fee is then included on save, so existing logics do not break. External apps, if required, can handle the fees during reconciliation.

This PR closes #50011 

Documentation: https://docs.frappe.io/erpnext/user/manual/en/bank-transaction

Please backport to v15

0xD0M1M0 committed:
feat: introduce extended bank transaction fields

0xD0M1M0 committed:
Merge branch 'develop' into feat-bank-transaction-fields

barredterra commented on erpnext/accounts/doctype/bank_transaction/bank_transaction.json:
```diff
@@ -234,12 +241,51 @@
+   "fieldtype": "Link",
+   "label": "Original Amount Currency",
+   "options": "Currency"
+  },
```
If fee and original amount can be in different currencies, we'd likely need two exchange_rate fields. Or add some validation that we have at most two currencies in a Bank Transaction record.

0xD0M1M0 replied:
Removed for now

barredterra commented on erpnext/accounts/doctype/bank_transaction/bank_transaction.json:
```diff
@@ -234,12 +241,51 @@
+  {
+   "fieldname": "included_fee",
+   "fieldtype": "Currency",
+   "label": "Included Fee",
```
If the fee is deemed included for withdrawals, it must likely be deemed excluded for deposits. Or does it just make sense to call it "included" in both cases? (We want the withdrawal/deposit amount to be the change in the bank balance. `deposit = receivable - fee`; `withdrawal = payable + fee`) 🤔

0xD0M1M0 replied:
Introduced included fee and excluded fee, that turns after save into an included fee. This way it does not break any existing logic

barredterra commented on erpnext/accounts/doctype/bank_transaction/bank_transaction.json:
```diff
@@ -234,12 +241,51 @@
    "fieldtype": "Column Break"
+  },
+  {
+   "fieldname": "included_fee",
```
Should we add a validation: Included fee cannot be more than withdrawal?

0xD0M1M0 replied:
The fee can be for a deposit or withdrawal. I am not against the validation.

What about an automatic Absolute on deposit and withdrawal so stupid bank imports, which cause random hick ups are eliminated. As far as I see, all logic depends on positive values anyway.

0xD0M1M0 commented:
I will continue working on it. Looking at different cases and excluded_fee as well.

0xD0M1M0 commented:
Still ongoing

0xD0M1M0 committed:
feat: reduce bank extended_bank_statement_section

0xD0M1M0 committed:
feat: Implement fee handling in bank transaction

Added handling for excluded fees in bank transactions.

0xD0M1M0 committed:
Merge branch 'develop' into feat-bank-transaction-fields

0xD0M1M0 committed:
fix: lost option

0xD0M1M0 committed:
Fix: JSON formatting for bank_party_iban label

0xD0M1M0 commented:
@barredterra I reduced the scope and implemented logic.

barredterra commented on erpnext/accounts/doctype/bank_transaction/bank_transaction.py:
```diff
@@ -307,6 +311,21 @@ def auto_set_party(self):
+	def handle_included_fee(self):
+		if self.included_fee and self.withdrawal:
+			if self.included_fee > self.withdrawal:
+				frappe.throw(_("Included fee is bigger than the withdrawal itself."))
```
Included fees can be charged for deposits as well, e.g. when receiving a foreign currency. Should we at least add a comment why this logic doesn't apply there?

barredterra commented on erpnext/accounts/doctype/bank_transaction/bank_transaction.py:
```diff
@@ -307,6 +311,21 @@ def auto_set_party(self):
+		# Include the excluded fee on validate to handle all further processing the same
+		if self.excluded_fee and self.excluded_fee > 0:
+			if self.deposit > 0:
+				self.deposit = self.deposit + self.excluded_fee
```
```suggestion
				self.deposit = self.deposit - self.excluded_fee
```

0xD0M1M0 committed:
fix: excluded fee handling for deposits

- Corrected the handling of excluded fees by subtracting them from deposits instead of adding.
- Added comment, why included_fees for deposits can be larger than the deposit itself.

0xD0M1M0 committed:
fix: linters

0xD0M1M0 committed:
fix: linters

0xD0M1M0 commented:
not stale

barredterra commented on erpnext/accounts/doctype/bank_transaction/bank_transaction.json:
```diff
@@ -233,12 +236,30 @@
+   "options": "currency"
+  },
+  {
+   "description": "On save, the fees will be included in the deposit and withdraw amounts, as well as in the included fee.",
```
```suggestion
   "description": "On save, the Excluded Fee will be converted to an Included Fee.",
```

I feel like this is easier to understand

barredterra commented on erpnext/accounts/doctype/bank_transaction/bank_transaction.py:
```diff
@@ -121,7 +125,7 @@ def before_submit(self):
 		self.set_status()
 
-		if frappe.get_single_value("Accounts Settings", "enable_party_matching"):
+		if frappe.db.get_single_value("Accounts Settings", "enable_party_matching"):
```
```suggestion
		if frappe.get_single_value("Accounts Settings", "enable_party_matching"):
```

Why should we remove caching here? Seems like an unrelated change.

0xD0M1M0 replied:
I couldn't find documentation on frappe.get_single_value and exchanged it for the document one

barredterra replied:
`frappe.get_single_value` uses the redis cache, `frappe.db.get_single_value` reads from the database every time.

barredterra commented on erpnext/accounts/doctype/bank_transaction/bank_transaction.py:
```diff
@@ -307,6 +311,26 @@ def auto_set_party(self):
+
+		if self.included_fee and self.withdrawal:
+			if self.included_fee > self.withdrawal:
+				frappe.throw(_("Included fee is bigger than the withdrawal itself."))
```
This should likely be called `validate_included_fee` and get called on `validate`, not `before_validate`.

barredterra commented on erpnext/accounts/doctype/bank_transaction/bank_transaction.py:
```diff
@@ -307,6 +311,26 @@ def auto_set_party(self):
+			if self.withdrawal > 0:
+				self.withdrawal = self.withdrawal + self.excluded_fee
+			self.included_fee = self.included_fee + self.excluded_fee
+			self.excluded_fee = 0
```
I think this nitpick from coderabbit makes sense, the code becomes more robust and understandable.

> - If both deposit and withdrawal are positive, the current if/if will adjust both, which is probably not intended if a Bank Transaction is conceptually one-directional.
> - Purely relying on truthiness and raw +/- can be brittle if any of these fields are None or stringy; using flt throughout keeps the math safe and consistent with the rest of the class.
>
> If domain-wise a transaction should never have both a non-zero deposit and withdrawal, and you’re okay to treat that as an invalid state, you could tighten the logic like this:

```suggestion
	def handle_excluded_fee(self):
		# Include the excluded fee on validate to handle all further processing the same
		excluded_fee = flt(self.excluded_fee)
		if excluded_fee <= 0:
			return

		# Enforce directionality
		if flt(self.deposit) > 0 and flt(self.withdrawal) > 0:
			frappe.throw(_("Only one of Deposit or Withdrawal should be non-zero when applying an excluded fee."))

		if flt(self.deposit) > 0:
			self.deposit = flt(self.deposit) - excluded_fee
		elif flt(self.withdrawal) > 0:
			self.withdrawal = flt(self.withdrawal) + excluded_fee

		self.included_fee = flt(self.included_fee) + excluded_fee
		self.excluded_fee = 0
```

> This keeps the behavior the same in the common cases while guarding against ambiguous transactions and type issues.

0xD0M1M0 committed:
chore: improve code stability

0xD0M1M0 committed:
fix: nitpick fix

0xD0M1M0 committed:
fix: add validation for empty deposit or withdrawal

0xD0M1M0 committed:
feat: added handling for negative deposits

0xD0M1M0 commented on erpnext/accounts/doctype/bank_transaction/bank_transaction.py:
```diff
@@ -307,6 +311,42 @@ def auto_set_party(self):
+
+	def handle_excluded_fee(self):
+		# Include the excluded fee on validate to handle all further processing the same
+		excluded_fee = flt(self.excluded_fee)
+		if excluded_fee <= 0:
+			return
+
+		# Suppress a negative deposit (aka withdrawal), likely not intendend
+		if flt(self.deposit) > 0 and (flt(self.deposit) - excluded_fee) < 0:
+			frappe.throw(
+				_("The Excluded Fee is bigger than the Deposit it is deducted from.")
+			)
+
+		# Enforce directionality
+		if flt(self.deposit) > 0 and flt(self.withdrawal) > 0:
+			frappe.throw(
+				_("Only one of Deposit or Withdrawal should be non-zero when applying an Excluded Fee.")
+			)
+
+		if flt(self.deposit) > 0:
+			self.deposit = flt(self.deposit) - excluded_fee
+		# A fee applied to deposit and withdrawal equal 0 become a withdrawal
+		elif flt(self.withdrawal) >= 0:
+			self.withdrawal = flt(self.withdrawal) + excluded_fee
+		self.included_fee = flt(self.included_fee) + excluded_fee
+		self.excluded_fee = 0
```
As the comment in the code states, a withdrawal creation is intended behavior. There cannot be an excluded fee, which is converted into an included fee without a negative effect on the account. This effect is a withdrawal.

0xD0M1M0 committed:
fix: linters

barredterra committed:
test(Bank Transaction): included/excluded fees

barredterra committed:
test(Bank Transaction): included/excluded fees (part 2)

barredterra committed:
style: break docstring

barredterra committed:
test: rename class to avoid conflicts
